### PR TITLE
feat(lsp): implement :lsp command

### PR DESCRIPTION
**Pull request by ii14**
All actual functionality is implemented inside `vim.lsp.command` module, which `:lsp` command simply `require`s. It's possible for plugins and users to either extend `:lsp` with more subcommands, or replace the module entirely and make something custom from scratch.

* `:lsp` and `:lsp!` is just `:LspInfo`, but without the floating window. Prints all active clients with `!`, or just clients attached to the current buffer. By default prints just attached buffers and root directory, the rest of properties are hidden behind `:verbose lsp`.
* `:lsp attach` attaches the buffer to a client, with format `42` or `clangd(42)`. Completion expands to the latter format, to make it easy to target clients by name.
* `:lsp detach` detaches the buffer from a client. Argument format is the same as above, but on top of that with no arguments will detach from all clients, and with just client name like `clangd` will detach from all clients matching that name.
* `:lsp definition`, `:lsp declaration`, `:lsp implementation`, `:lsp typedefinition`, `:lsp references` etc. are nothing fancy, they just call their respective `vim.lsp.buf` functions.
* `:lsp rename` can take new symbol name as an argument.
* `:lsp find` is `workspace/symbol`, can take in symbol name as an argument.
* `:lsp format` for now can take arguments in format `async` and `timeout=1000`. Also works with range, and in that case those arguments are not allowed.
* `:lsp codeaction` for now can take a single argument in format `only=foo,bar`, to request only specific code actions kinds.

In the standard vim command fashion, `:lsp` subcommand names can be matched by giving just their substring, eg. `:lsp f` will run `:lsp format`, `:lsp c` -> `:lsp codeaction` etc.

Command line completion results depend on the context. For example, if none of the attached clients support `typeDefinitionProvider`, `typedefinition` will not be included in completion. If `:lsp` is prefixed with range, only subcommands that accept ranges will be suggested (namely `format` and `codeaction`).

To add a new subcommand, users can just append them to the `require('vim.lsp.command').commands` table. Each subcommand is a table with the following properties:
* `name` - string, self explanatory.
* `run` - function, receives context table as a single parameter. Inside the table there is a buffer handle, optional arguments to the subcommand, range, `:verbose` modifier, `!`.
* `check` - function, optional. It's used to do validation before running the command, but also to figure out if the subcommand should be included in completion. Receives context table as a single parameter, should return a string in case of error, or `nil` on success.
* `expand` - function, optional. Used for completion. Receives completed argument as the first parameter, context as the second. Expects back a list of completion items, and optionally the starting position offset.

### CODE REVIEW

**Comment by gpanders** [comment]
I haven't looked closely at the implementation, but in general I'm in favor of a `:lsp` command. At the very least, this would allow us to upstream `:LspStart`, `:LspStop`, and `:LspInfo` from nvim-lspconfig.

LSP has a lot of functionality and I, personally, only create key mappings for the functions I use most often. Having an easy to use `:lsp` command for the less often used (but still useful) functions seems useful to me.

This could also be a good extension point for plugins to hook into and provide additional functionality.

**Comment by clason** [comment]
...as well as exposing custom commands that some servers provide.

File: src/nvim/lua/executor.c [hunk 1]
```diff
@@ -1909,3 +1910,186 @@ void nlua_do_ucmd(ucmd_T *cmd, exarg_T *eap)
...
+    emsg(_("vim.lsp.command is not a table"));
+    return;
+  }
+
+  lua_getfield(lstate, -1, "run");
+  if (lua_type(lstate, -1) != LUA_TFUNCTION) {
+    lua_pop(lstate, 1);
+    emsg(_("vim.lsp.command.run is not a function"));
+    return;
+  }
```
Comment (bfredl) [inline src/nvim/lua/executor.c:1941]: we could load the string `"require('vim.lsp.command').run(...)"` instead and have luajit generate all errors for us, instead of the hand-coded prelude.

**Comment by mfussenegger** [comment]
I'm kinda opposed to this, at least for now.


The main advantage this would bring to users is to type a little less if they didn't setup any keymaps. Looking at parts of the implementation I think the additional maintenance cost is quite high for that.

The other aspect is that I'm not sure if this improves the interface. Products with many knobs are usually harder to use than products with fewer knobs. That's okay if those extra knobs increase the utility of a product, not so much if they're just duplicate.


Now onto some more specifics:

- `:lsp` / `:LspInfo`: LspInfo currently does more than just printing all clients, it also prints all *not* started yet. It can do that because lsp-config currently includes the "managing clients" functionality. I think the first step we have to do to bring anything like that into core is to move the manager component.


- `:lsp find` would have overlap with the functionality one gains by using the lsp tagfunc with all of `:tselect` and so on. I think we should consider having a deeper integration of lsp features with other existing neovim features and improve those if necessary, instead of treating lsp as kind of a separate island. Similar how it has already happened with vim.diagnostic. Instead of creating more dedicated lsp entry-points, have "first class" entry points where lsp is just *one* possible provider. 

- `:lsp format`: Related to the previous point. In addition to vim.lsp.buf.format there is also a formatexpr implementation to get integration with `gq` and so on. How many ways to format a buffer do we really need?

- Some of the existing functions still have issues if used with multiple clients. This may still require some interface changes, similar to how it happend with `format`. Adding a second interface now might make that work more troublesome and I'd rather avoid that.

**Comment by mfussenegger** [comment]
Some more thoughts/ideas on this:


- We could consider moving the functions that are currently in the `vim.lsp.buf` namespace to `vim.lsp`

- Maybe the completion for `:lua` could be extended to also support function arguments? At least for built-in functions where we could generate some metadata based on the lua docs.

- Maybe there could be some kind of command abbreviation feature, so that `:lsp<tab>` expands to `:lua vim.lsp.`, or that `:lsp` is just an alias for `:lua vim.lsp.` and you can call functions like `:lsp.rename { ... }`

**Comment by noahfrederick** [comment]
I have difficulty with the UX introduced here:

1. I'm not aware of a precedent for a built-in command that takes sub-command arguments like this, and not in favor of introducing new conventions unless absolutely necessary. Caveats such as "command x *sometimes* accepts a range" is indicative of this not being a desirable pattern here.
2. `:lsp` conflates all sorts of functionality that is related in terms of the implementation (via LSP), but not really related from a user's perspective, especially management of LSP clients vs. editing commands. I'm strongly in favor of enhancing existing mechanisms and idioms, as @mfussenegger already gave examples of, instead of adding another group of commands that overlap with existing Neovim conventions.
3. The name `:lsp` itself is an *abbreviation* of an *implementation detail*, which seems perverse. But I would much rather have a set of top-level commands (where warranted) that are descriptive of what they do, which sidesteps the problem of coming up with a better name.

**Comment by ii14** [comment]
> I'm not aware of a precedent for a built-in command that takes sub-command arguments like this

There is `:highlight`, `:syntax`, `:csope` for example. The last one in particular is basically what this PR would do, but for cscope.

**Comment by noahfrederick** [comment]
> There is `:highlight`, `:syntax`, `:cscope` for example.

Ah, yeah, `:cscope` definitely uses this convention in the same way. I stand corrected. For what it's worth, I would have made the exact same comments would we have been talking about introducing `:cscope`.

> vim features are already named after implementation details like that, like the mentioned cscope or ctags.

Absolutely—Vim is full of unintuitively named commands and options of all sorts.

Because of that design decision we're now stuck with a `:cscope` command that potentially overlaps with a bunch of LSP functionality, but its interface can't be reused, because the interface is cscope-specific.

**Comment by andmis** [comment]
I'm a new contributor, but my 2 cents:

I feel like one of the major strengths of Neovim over Vim is having Lua as the first-class interface as much as possible:
- You can `vim.keymap.set` keystokes to Lua functions.
- You can call Lua functions interactively with `:lua`.
- You can trivially call Lua functions _from other Lua functions_, making ad-hoc composition easy.
- The syntax and semantics of Lua functions are familiar to everyone -- Ex commands are a completely different beast, and the syntax for composing them is arcane.
- The `:map <Cmd>MyExCommandString<CR>` syntax is arcane and user-unfriendly.

The last point is maybe the important one -- many plugins still expose their functionality primarily via Ex commands (or `<Plug>` commands). Part of the reason is that people don't like to write or interact with Vim script so Vim script functions never became the first-class interface for Vim functionality. But Lua functions are great! And Lua is a great language. So we should push towards making Lua functions the first-class interface.

In other words, I can see the use case for having commands like `:lsp attach ...`, that sounds very convenient as an interactive interface. But IMHO these commands should be implemented and treated as thin quality-of-life wrappers around Lua functions providing all of the main functionality, and user-friendliness at the Lua interface should be the top priority.

**Comment by ii14** [comment]
> its interface can't be reused, because the interface is cscope-specific.

Well yes, but that's because its behavior is hardcoded in C (I assume). This is not, the actual implementation is in lua. You could just replace the entire lua module with your own thing, provide `run` and `expand` functions and define your own interface.

**Comment by ii14** [comment]
> But IMHO these commands should be implemented and treated as thin quality-of-life wrappers around Lua functions providing all of the main functionality, and user-friendliness at the Lua interface should be the top priority.

They are, that's pretty much the purpose of this PR. This is by no means meant to replace the existing interface. You won't even be able to do some of the more advanced stuff you can do with lua. This is just to have a quick access to basic functions without having to write lua, which in my opinion is not the most convenient language to use on a command line.

**Comment by justinmk** [comment]
Thanks for your help here! I long thought about repurposing `:cscope` for this...

> At the very least, this would allow us to upstream `:LspStart`, `:LspStop`, and `:LspInfo` from nvim-lspconfig.

We might need `:lsp stop/start`, but the extensible `commands` thing looks like an unnecessary inner platform. We already have a plugin system; coordinating another namespace or registry looks like a lot of noise with little gain.

I also strongly agree with the comments from @mfussenegger , @andmis :

> Maybe the completion for `:lua` could be extended
> ...
> one of the major strengths of Neovim over Vim is having Lua as the first-class interface

Commands only exist as an ad-hoc UI, to avoid needing to quote string args. Not sure how common LSP tasks are that wouldn't instead be a mapping.

## Proposal

- `:checkhealth lsp` should do what `:LspInfo` does.
- We need to offer strong default mappings for LSP.
    - `:help lsp-config` used to have mappings that could be pasted-into a user config, but that was removed :( 
    - `lsp.setup()` (or whatever) should set things like `ctrl-<space>`, `gd`, `gr` (find all references), etc. 
        - User can opt-out with something like `lsp.setup({default_mappings=false})`
